### PR TITLE
Add disown() method to kj::Own

### PR DESCRIPTION
**Pull request by garrettgu10**
Addresses #2089

### CODE REVIEW

File: c++/src/kj/memory.h [hunk 1]
```diff
@@ -313,6 +313,14 @@ class Own<T, decltype(nullptr)> {
   inline operator T*() { return ptr; }
   inline operator const T*() const { return ptr; }
 
+  // Surrenders ownership of the underlying object to the caller. The caller must pass in the 
+  // correct disposer to prove that they know how the object is meant to be disposed of. 
+  inline T* disown(const Disposer* d) {
+    KJ_REQUIRE(d == disposer, "When disowning an object, disposer must be equal to Own's disposer");
```
Comment (kentonv) [inline c++/src/kj/memory.h:319]: KJ_REQUIRE isn't defined here, and you probably can't #include it without creating a cyclic dependency. You will need to do something like:

`if (d != disposer) throwWrongDisposerError();`

and then implement `throwWrongDisposerError()` out-of-line (in `memory.c++`).

File: c++/src/kj/memory.h [hunk 2]
```diff
@@ -224,6 +224,8 @@ using PtrCounter = AtomicPtrCounter;
 template <typename T, typename StaticDisposer = decltype(nullptr)>
 class Own;
 
+void throwWrongDisposerError();
```
Comment (kentonv) [inline c++/src/kj/memory.h:227]: Nitpick: Probably should be in `_` namespace, where we put everything that's not intended to be a public API...

```
namespace _ {  // (private)
void throwWrongDisposerError();
}  // namespace _ (private)
```

**Comment by jasnell** [comment]
The change generally LGTM but this needs tests

**Comment by kentonv** [comment]
Sorry for the back-and-forth but I agree with James, there should be a unit test.

File: c++/src/kj/memory-test.c++ [hunk 3]
```diff
@@ -370,6 +370,20 @@ TEST(Memory, OwnConstVoid) {
     maybe = kj::none;
     KJ_EXPECT(maybe == kj::none);
   }
+
+  {
+    bool destructorCalled = false;
+    Own<SingularDerivedDynamic> ptr = heap<SingularDerivedDynamic>(123, destructorCalled);
+    SingularDerivedDynamic* addr = ptr.get();
+
+    ptr.disown(&_::HeapDisposer<SingularDerivedDynamic>::instance);
```
Comment (kentonv) [inline c++/src/kj/memory-test.c++:379]:
```suggestion
    KJ_EXPECT(ptr.disown(&_::HeapDisposer<SingularDerivedDynamic>::instance) == addr);
```

**Review by kentonv — APPROVED** [review]
Will merge when CI finishes.

(If someone else merges, please be sure to squash-merge...)

File: c++/src/kj/memory.h [hunk 4]
```diff
@@ -429,6 +440,16 @@ class Own {
   inline operator T*() { return ptr; }
   inline operator const T*() const { return ptr; }
 
+  // Surrenders ownership of the underlying object to the caller. The caller must pass in the 
+  // correct disposer to prove that they know how the object is meant to be disposed of. 
+  template<typename SD>
+  inline T* disown() {
+    static_assert(std::is_same<StaticDisposer, SD>::value, "disposer must be the same as Own's disposer");
```
Comment (kentonv) [inline c++/src/kj/memory.h:447]:
```suggestion
    static_assert(kj::isSameType<StaticDisposer, SD>, "disposer must be the same as Own's disposer");
```

**Pull request merged**